### PR TITLE
First stab at job sets with htcondor CLI tool HTCONDOR-793

### PR DESCRIPTION
**Pull request by jasoncpatton**
Added job sets to `htcondor` CLI tool with noun-verbs:
`jobset submit`
`jobset list`
`jobset status [job_set_name]`
`jobset remove [job_set_name]`

Job set submit files are mostly parsed "by hand",
using `htcondor.Submit()` to parse job submit syntax.
Job set jobs are denoted by having `JobSetName` and `InJobSet`
defined in their Job ClassAds.

Marking as a draft pull request as docs are probably needed, and there definitely could be some UI tweaks.

Example output:

```
(cli_test) [jcpatton@submit-1 cli_test]$ cat test_job_set.setsub
name = StateComparisonJobs
# Type of iterator "table" followed by iterator inputs args
iterator = table A,B {
    # contents of table.txt
    wisconsin,iowa
    iowa,illinois
    illinois,wisconsin
}
job {
  # The syntax here is EXACTLY like a job submit file,
  # all the same macros work... just no need for a "queue" statement.
  executable = print_and_sleep.sh
  arguments = $(A).data $(B).data
  log = first_job.log
  output = $(A)_$(B).first_job.out
  error = $(A)_$(B).first_job.err
  request_disk = 20MB
  request_cpus = 1
  request_memory = 20MB
}
job C=A D=B {
  executable = print_and_sleep.sh
  arguments = $(C).data $(D).data
  log = second_job.log
  output = $(C)_$(D).second_job.out
  error = $(C)_$(D).second_job.err
  request_disk = 20MB
  request_cpus = 1
  request_memory = 20MB
}
(cli_test) [jcpatton@submit-1 cli_test]$ ./htcondor jobset submit test_job_set.setsub
Submitted job set StateComparisonJobs containing 2 jobs.
(cli_test) [jcpatton@submit-1 cli_test]$ ./htcondor jobset list
JOB_SET_NAME
StateComparisonJobs
(cli_test) [jcpatton@submit-1 cli_test]$ ./htcondor jobset list --allusers
OWNER     JOB_SET_NAME
jcpatton  StateComparisonJobs
(cli_test) [jcpatton@submit-1 cli_test]$ ./htcondor jobset status StateComparisonJobs

-- Schedd: submit-1.chtc.wisc.edu : <128.105.244.191:9618?... @ 10/26/21 14:34:45
BATCH_NAME                  SUBMITTED  DONE  RUN  IDLE  TOTAL  JOB_IDS              
Set: StateComparisonJobs  10/26 14:34     -    -     6      6  14362513.0-14362514.2
6 jobs; 0 completed; 0 removed; 6 idle; 0 running; 0 held; 0 suspended

(cli_test) [jcpatton@submit-1 cli_test]$ ./htcondor jobset status StateComparisonJobs --nobatch

-- Schedd: submit-1.chtc.wisc.edu : <128.105.244.191:9618?... @ 10/26/21 14:34:56
BATCH_NAME                  SUBMITTED  DONE  RUN  IDLE  TOTAL  JOB_IDS              
ID: 14362513              10/26 14:34     -    -     3      3  14362513.0-2         
ID: 14362514              10/26 14:34     -    -     3      3  14362514.0-2         
------------------------------------------------------------------------------------
Set: StateComparisonJobs  10/26 14:34     -    -     6      6  14362513.0-14362514.2
6 jobs; 0 completed; 0 removed; 6 idle; 0 running; 0 held; 0 suspended

(cli_test) [jcpatton@submit-1 cli_test]$ ./htcondor jobset remove StateComparisonJobs
Removed 6 jobs matching job set StateComparisonJobs for user jcpatton.
(cli_test) [jcpatton@submit-1 cli_test]$ ./htcondor jobset list
No activate job sets found.
(cli_test) [jcpatton@submit-1 cli_test]$ ./htcondor jobset status StateComparisonJobs
No active job sets found matching "StateComparisonJobs" for user jcpatton.
```

### CODE REVIEW

File: src/condor_tools/htcondor_cli/job_set.py [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+        },
+    }
+
+    def __init__(self, logger, job_set_file, **options):
+
+        self.name = None
+        self.itemdata = None
+        self.jobs = []
+
+        schedd = htcondor.Schedd()
```
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_tools/htcondor_cli/job_set.py:33]: Noted for the future, we'll update this when 811 gets resolved.

File: src/condor_tools/htcondor_cli/job_set.py [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+                line = f.readline()
+                if line == "":
+                    break
+                lineno += 1
+
+                line = line.strip()
+                if line == "" or line.startswith("#"):
+                    continue
+
+                command = line.split()[0].split("=")[0].casefold()
```
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_tools/htcondor_cli/job_set.py:87]: I'm curious how you got this one to trip... I'll catch an `IndexError` here anyway.

File: src/condor_tools/htcondor_cli/job_set.py [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+                            inline = "{"
+                            inlineno = 0
+                            inline_data = ""
+                            while inline != "":
+                                inline = f.readline()
+                                inlineno += 1
+                                if inline.split("#")[0].strip() == "}":
+                                    break
+                                inline_data += inline
+                            else:
```
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_tools/htcondor_cli/job_set.py:128]: I was able to get this to trip in testing... ah, but if it finds another `}` later (like in a `job`)... this is a toughie, I'll try to put together something that catches this situation.

File: src/condor_tools/htcondor_cli/job_set.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_tools/htcondor_cli/job_set.py]: Resolved in aaa3d51bfa9dc60be00242d183d0f61efebf4229

File: src/condor_tools/htcondor_cli/job_set.py [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+                    if self.name is not None:
+                        raise ValueError(f"""Job set name can only be set once, second name found in {job_set_file} at line {lineno}.""")
+                    value = line.split("#")[0].split("=")[1].strip()
+
+                    self.name = value
+
+                elif command == "iterator":
+                    if self.itemdata is not None:
+                        raise ValueError(f"""Job set iterator can only be set once, second iterator found in {job_set_file} at line {lineno}.""")
+                    value = line.split("#")[0].split("=")[1].strip()
```
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_tools/htcondor_cli/job_set.py]: Added in aaa3d51bfa9dc60be00242d183d0f61efebf4229

File: src/condor_tools/htcondor_cli/job_set.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_tools/htcondor_cli/job_set.py:154]: Added a check for new `{` and throw error with line number in aaa3d51bfa9dc60be00242d183d0f61efebf4229

File: src/condor_tools/htcondor_cli/job_set.py [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+                        inline = "{"
+                        inlineno = 0
+                        inline_data = ""
+                        while inline != "":
+                            inline = f.readline()
+                            inlineno += 1
+                            if inline.split("#")[0].strip() == "}":
+                                break
+                            inline_data += inline.lstrip()
+                        else:
```
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_tools/htcondor_cli/job_set.py:210]: Added a check for new `{` and throw error with line number in aaa3d51bfa9dc60be00242d183d0f61efebf4229

File: src/condor_tools/htcondor_cli/job_set.py [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+                    continue
+
+                command = line.split()[0].split("=")[0].casefold()
+                if command not in commands:
+                    raise ValueError(f"""Unrecognized command "{command}" in {job_set_file} at line {lineno}.""")
+
+                if command == "name":
+                    if self.name is not None:
+                        raise ValueError(f"""Job set name can only be set once, second name found in {job_set_file} at line {lineno}.""")
+                    value = line.split("#")[0].split("=")[1].strip()
```
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_tools/htcondor_cli/job_set.py]: Added in aaa3d51bfa9dc60be00242d183d0f61efebf4229

File: src/condor_tools/htcondor_cli/job_set.py [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+                                self.itemdata = self.parse_columnar_itemdata(iterator_names, f_iter.read(), fname=iterator_source)
+
+                elif command == "job":
+                    value = " ".join(line.split("#")[0].strip().split()[1:])
+
+                    # Get the variable name mappings
+                    mapping_strs = value.split()[:-1]
+                    mappings = []
+                    for mapping_str in mapping_strs:
+                        mappings.append(tuple(x.strip() for x in mapping_str.split("=")))
```
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_tools/htcondor_cli/job_set.py]: This was overhauled in aaa3d51bfa9dc60be00242d183d0f61efebf4229, there are checks for if a mapping string even exists, if `=` is in the mapping string, and if `len(mapping_str.split("=")) != 2`, with appropriate errors thrown.

File: src/condor_tools/htcondor_cli/job_set.py [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+                            inlineno += 1
+                            if inline.split("#")[0].strip() == "}":
+                                break
+                            inline_data += inline.lstrip()
+                        else:
+                            raise ValueError(f"""Unclosed bracket in {job_set_file} starting at line {lineno}.""")
+                        lineno += inlineno
+                        submit_obj = htcondor.Submit(inline_data)
+                    else:
+                        with open(job_source, "rt") as f_sub:
```
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_tools/htcondor_cli/job_set.py]: Fixed by doing a proper `value.split()` earlier in the code block in aaa3d51bfa9dc60be00242d183d0f61efebf4229

File: src/condor_tools/htcondor_cli/job_set.py [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+                    # Remap variables in the Submit object
+                    submit_obj = self.remap_submit_variables(mappings, submit_obj)
+
+                    # Store each job
+                    self.jobs.append(submit_obj)
+
+        # Add job set name to each job's Submit object
+        for i_job, job in enumerate(self.jobs):
+            job["MY.JobSetName"] = classad.quote(self.name)
+            job["MY.InJobSet"] = True
```
[the remark this answers is not in the document]
Comment (jasoncpatton) [reply src/condor_tools/htcondor_cli/job_set.py:230]: Resolved per Slack convo

**Review by markcoatsworth — APPROVED** [review]

**Comment by markcoatsworth** [comment]
LGTM. We have some work to do in follow-up tickets but this is good for now.

**Pull request merged**